File: app/src/main/java/com/aotem/hg51e/MainActivity.java
package com.aotem.hg51e;

import androidx.appcompat.app.AppCompatActivity;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import com.aotem.hg51e.net.NetworkUtils;
import com.hjq.shape.view.ShapeButton;
import com.hjq.shape.view.ShapeEditText;
import com.hjq.toast.Toaster;

import io.reactivex.rxjava3.annotations.NonNull;
import io.reactivex.rxjava3.observers.DisposableObserver;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {
    private static final String TAG = "MainActivity";
    private ShapeEditText et_account;
    private ShapeEditText et_password;
    private ShapeEditText et_host;
    private ShapeButton btn_get;
    private ShapeButton btn_copy;

    private ClipboardManager clipboardManager;
    private ConnectivityManager connectivityManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        clipboardManager = (ClipboardManager) getSystemService(CLIPBOARD_SERVICE);
        connectivityManager = (ConnectivityManager) getSystemService(CONNECTIVITY_SERVICE);
        et_account = findViewById(R.id.et_account);
        et_password = findViewById(R.id.et_password);
        et_host = findViewById(R.id.et_host);
        btn_get = findViewById(R.id.btn_get);
        btn_copy = findViewById(R.id.btn_copy);
        btn_copy.setOnClickListener(this);
        btn_get.setOnClickListener(this);

    }

    @Override
    public void onClick(View v) {
        if (v.getId() == R.id.btn_get) {
            String host = et_host.getText().toString();
            if (host.isEmpty()) {
                host = "192.168.1.1";
            }
            Network network = connectivityManager.getActiveNetwork();
            NetworkCapabilities networkCapabilities = connectivityManager.getNetworkCapabilities(network);
            if (networkCapabilities != null) {
                if (networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_VPN)) {
                    Toaster.showShort("请关闭你的代理");
                    return;
                } else if (networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)) {
                    NetworkUtils.queryUser("http://" + host).safeSubscribe(new DisposableObserver<User>() {
                        @Override
                        public void onNext(@NonNull User user) {
                            Log.d(TAG, "onNext: ");
                            Toaster.showShort("获取成功");
                            String account = user.getData().getWeb_tele_account_name();
                            String password = user.getData().getTeleAccount_password();
//                            Log.d(TAG, "onNext: " + account);
//                            Log.d(TAG, "onNext: " + password);
                            et_account.setText(account);
                            et_password.setText(password);

                        }

                        @Override
                        public void onError(@NonNull Throwable e) {
                            Toaster.showShort("获取失败");
//                            Log.d(TAG, "onError: " + e.getMessage());
                        }

                        @Override
                        public void onComplete() {

//                            Log.d(TAG, "onComplete: ");
                        }
                    });
                } else if (networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)) {
                    Toaster.showShort("请连接上无线网络");
                }
            } else {
                Toaster.showShort("当前无网络连接");
            }

        } else if (v.getId() == R.id.btn_copy) {
            String password = et_password.getText().toString();
            if (password.isEmpty()) {
                Toaster.showShort("当前密码为空，请先获取密码");
                return;
            }
            if (clipboardManager != null) {
                clipboardManager.setPrimaryClip(ClipData.newPlainText("password", password));
                Toaster.showShort("已复制到剪切板");
            }
        }
    }
}

File: app/src/main/java/com/aotem/hg51e/MainActivity.kt
package com.aotem.hg51e

import android.content.ClipData
import android.content.ClipDescription
import android.content.ClipboardManager
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import android.os.Bundle
import android.os.PersistableBundle
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.aotem.hg51e.data.remote.HostUrl
import com.aotem.hg51e.ui.main.MainViewModel
import com.aotem.hg51e.ui.main.UiState
import com.hjq.shape.view.ShapeButton
import com.hjq.shape.view.ShapeEditText
import com.hjq.toast.Toaster
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

@AndroidEntryPoint
class MainActivity : AppCompatActivity() {

    private val viewModel: MainViewModel by viewModels()

    private lateinit var etAccount: ShapeEditText
    private lateinit var etPassword: ShapeEditText
    private lateinit var etHost: ShapeEditText
    private lateinit var btnGet: ShapeButton
    private lateinit var btnCopy: ShapeButton

    private val clipboardManager: ClipboardManager by lazy {
        getSystemService(CLIPBOARD_SERVICE) as ClipboardManager
    }
    private val connectivityManager: ConnectivityManager by lazy {
        getSystemService(CONNECTIVITY_SERVICE) as ConnectivityManager
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        etAccount = findViewById(R.id.et_account)
        etPassword = findViewById(R.id.et_password)
        etHost = findViewById(R.id.et_host)
        btnGet = findViewById(R.id.btn_get)
        btnCopy = findViewById(R.id.btn_copy)

        btnGet.setOnClickListener { onGetClicked() }
        btnCopy.setOnClickListener { onCopyClicked() }

        observeUiState()
        observeMessages()
    }

    private fun observeUiState() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.uiState.collect { state ->
                    when (state) {
                        is UiState.Idle -> Unit
                        is UiState.Loading -> {
                            btnGet.isEnabled = false
                            btnCopy.isEnabled = false
                        }
                        is UiState.Success -> {
                            btnGet.isEnabled = true
                            btnCopy.isEnabled = true
                            etAccount.setText(state.user.data.webTeleAccountName)
                            etPassword.setText(state.user.data.teleAccountPassword)
                        }
                        is UiState.Error -> {
                            btnGet.isEnabled = true
                            btnCopy.isEnabled = true
                        }
                    }
                }
            }
        }
    }

    private fun observeMessages() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.messages.collect { Toaster.showShort(it) }
            }
        }
    }

    private fun onGetClicked() {
        if (!btnGet.isEnabled) return

        val baseUrl = HostUrl.normalize(etHost.text?.toString())
        if (baseUrl == null) {
            Toaster.showShort("地址格式不正确")
            return
        }

        val capabilities = connectivityManager.activeNetwork
            ?.let { connectivityManager.getNetworkCapabilities(it) }
        if (capabilities == null) {
            Toaster.showShort("当前无网络连接")
            return
        }

        when {
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_VPN) ->
                Toaster.showShort("请关闭你的代理")
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) ||
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) ||
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) ->
                viewModel.queryUser(baseUrl.toString())
            else ->
                Toaster.showShort("当前网络类型不支持")
        }
    }

    private fun onCopyClicked() {
        val password = etPassword.text?.toString().orEmpty()
        if (password.isEmpty()) {
            Toaster.showShort("当前密码为空，请先获取密码")
            return
        }
        val clip = ClipData.newPlainText("password", password)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            clip.description.extras = PersistableBundle().apply {
                putBoolean(ClipDescription.EXTRA_IS_SENSITIVE, true)
            }
        }
        clipboardManager.setPrimaryClip(clip)
        Toaster.showShort("已复制到剪贴板")
    }
}
